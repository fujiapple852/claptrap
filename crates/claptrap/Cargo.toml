[package]
name = "claptrap"
description.workspace = true
version.workspace = true
authors.workspace = true
documentation.workspace = true
homepage.workspace = true
repository.workspace = true
keywords.workspace = true
categories.workspace = true
readme = "README.md"
license.workspace = true
edition.workspace = true
rust-version.workspace = true

[dependencies]
claptrap-core.workspace = true
anstream.workspace = true
anyhow.workspace = true
clap = { workspace = true, features = ["derive", "string", "cargo", "wrap_help", "usage", "unstable-styles", "color", "suggestions", "error-context", "env"] }
clap-markdown.workspace = true
clap_complete.workspace = true
clap_mangen.workspace = true
minijinja.workspace = true
schemars = { workspace = true, features = ["derive", "indexmap2"] }
serde_json = { workspace = true, features = ["preserve_order"] }
serde_yaml = { workspace = true, features = [] }
toml = { workspace = true, features = ["preserve_order"] }

[dev-dependencies]
insta = "1.43.1"
tempfile = "3.21.0"
test-case = "3.3.1"

[lints]
workspace = true

[package.metadata.generate-rpm]
assets = [
  { source = "target/release/claptrap", dest = "/usr/bin/claptrap", mode = "755" },
]
