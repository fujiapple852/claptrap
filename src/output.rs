use clap::builder::StyledStr;
use itertools::Itertools;

fn shell_quote(value: &str) -> String {
    format!("'{}'", value.replace('\'', "'\\''"))
}
use std::fmt::Display;

// The prefix for variables output by claptrap
const PREFIX: &str = "claptrap";

/// Represents the output of a claptrap command.
#[derive(Debug, Eq, PartialEq)]
pub enum Output {
    Cat(CatCmd),
    Variables(Vec<Var>),
}

impl Display for Output {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Cat(cmd) => write!(f, "{cmd}"),
            Self::Variables(vars) => write!(f, "{}", vars.iter().format("\n")),
        }
    }
}

/// Represents a variable output by claptrap.
#[derive(Debug, Eq, PartialEq)]
pub enum Var {
    Single(Vec<String>, String, String),
    Many(Vec<String>, String, Vec<String>),
}

impl Display for Var {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Single(prefix, name, value) => {
                let value = shell_quote(value);
                if prefix.is_empty() {
                    write!(f, "{PREFIX}_{name}={value}")
                } else {
                    write!(
                        f,
                        "{}_{}_{}={}",
                        PREFIX,
                        prefix.iter().format("_"),
                        name,
                        value
                    )
                }
            }
            Self::Many(prefix, name, values) => {
                let values = values.iter().map(|v| shell_quote(v)).join(" ");
                if prefix.is_empty() {
                    write!(f, "{PREFIX}_{name}=({values})")
                } else {
                    write!(
                        f,
                        "{}_{}_{}=({})",
                        PREFIX,
                        prefix.iter().format("_"),
                        name,
                        values
                    )
                }
            }
        }
    }
}

/// Exit code for the `CatCmd`.
#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ExitCode {
    Success = 0,
    Error = 1,
    Usage = 2,
    Panic = 3,
}

impl Display for ExitCode {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}", *self as i32)
    }
}

/// A `cat` command invocation.
#[derive(Debug, Eq, PartialEq)]
pub struct CatCmd {
    pub data: StyledStr,
    pub exit_code: ExitCode,
}

impl CatCmd {
    #[must_use]
    pub fn new(cmd: StyledStr, exit_code: ExitCode) -> Self {
        Self {
            data: cmd,
            exit_code,
        }
    }
}

impl Display for CatCmd {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "command cat <<'EOF'\n{}EOF\nexit {}",
            self.data.ansi(),
            self.exit_code
        )
    }
}
