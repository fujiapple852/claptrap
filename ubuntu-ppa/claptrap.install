target/release/claptrap usr/bin
