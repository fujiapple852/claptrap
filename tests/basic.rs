#[path = "util.rs"]
mod util;
use claptrap::output::{CatCmd, ExitCode};
use test_case::test_case;

#[test_case(case!("basic", "no_args"))]
#[test_case(case!("basic", "short_help"))]
#[test_case(case!("basic", "long_help"))]
#[test_case(case!("basic", "short_version"))]
#[test_case(case!("basic", "variables"))]
#[test_case(case!("basic", "unexpected_arg"))]
fn test_values((name, spec, args): (&str, &str, &str)) {
    insta::assert_snapshot!(name, util::run(spec, args));
}

#[test]
fn cat_cmd_handles_eof_in_message() {
    let styled = clap::builder::StyledStr::from("this contains EOF in the text\nEOF\nand more\n");
    let cmd = CatCmd::new(styled, ExitCode::Error);
    insta::assert_snapshot!(format!("{}", cmd));
}
