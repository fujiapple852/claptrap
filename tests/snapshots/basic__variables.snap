---
source: tests/basic.rs
expression: "util::run(spec, args)"
---
claptrap_mode='stream'
claptrap_protocol='udp'
